### DISCUSSION TITLE
Adjusting mp3 file output sample rate and dynamic range

### DISCUSSION THREAD

**Discussion by ejackson** [rtl-airband/RTLSDR-Airband#360] [Q&A]
Hello,

  I would like to improve the quality of the .mp3 file outputs from airband.  Currently the sample rate is 8kHz and the output volume is is really low.  I cannot find the settings to adjust these, and any help will be appreciated.

Thank you,

  Edmund

---------------
My config file looks like this:

# File outputting listener based on RTLSDR.

devices:
({
  type = "rtlsdr";
  index = 0;
  gain = 40;
  centerfreq = 462.500;
  afc = 1;
  channels:
  (
    {
      freq = 462.5625;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "1";
	  split_on_transmission = true;
        }
      );
    },

... etc

**Comment by charlie-foxtrot** [comment]
If you compile in `NFM` support the output audio will be 16 kHz, but this may not actually improve the audio quality any.

Doing a higher audio bitrate will only preserve audio quality.  The AM demodulator runs at 8 kHz and the NFM demodulator runs at 16 kHz.

As for turning up the volume, that should be a straight forward code change, I'll add that to the backlog

**Comment by ejackson** [comment]
Thank you.

16k vs 8k:  I beileve I compiled on unstable, with NFM enabled and still only getting 8k.  I've inlined a log below.  Is this a bug ?

Volume: if its a quick-fix I'd be happy to take a crack at it.

Thanks,

  Edmund

***

$ git checkout unstable
branch 'unstable' set up to track 'origin/unstable'.
Switched to a new branch 'unstable'

$ git pull
Already up to date.

$ cmake ../
-- Build type not specified: defaulting to Release
-- Found lame includes:	/usr/include/lame/lame.h
-- Found lame library: /usr/lib/x86_64-linux-gnu/libmp3lame.so
-- Checking for SoapySDR
--   SoapySDR found, /usr/include, SoapySDR
-- RTLSDR-Airband configuration summary:

-- - Build type:		Release
-- - Operating system:		Linux
-- - SDR drivers:
--   - librtlsdr:		requested: ON, enabled: TRUE
--   - mirisdr:			requested: OFF, enabled: FALSE
--   - soapysdr:		requested: ON, enabled: TRUE
-- - Other options:
--   - Platform:		default
--   - Broadcom VideoCore GPU:	FALSE
**--   - NFM support:		ON**
--   - PulseAudio:		requested: ON, enabled: TRUE
--   - Profiling:		requested: OFF, enabled: FALSE
--   - Icecast TLS support:	TRUE
-- Configuring done
-- Generating done
-- Build files have been written to: /home/edmund/Code/RTLSDR-Airband

$ cd..
$ make
$ sudo make install

$ rtl_airband -v
**RTLSDR-Airband version 4.0.3-3ac8ef4**

$ rtl_airband -f -c 

$ mplayer 4_20230303_171831.mp3 
MPlayer 1.4 (Debian), built with gcc-12 (C) 2000-2019 MPlayer Team
do_connect: could not connect to socket
connect: No such file or directory
Failed to open LIRC support. You will not be able to use your remote control.

Playing 4_20230303_171831.mp3.
libavformat version 59.27.100 (external)
Audio only file format detected.
Load subtitles in ./

Opening audio decoder: [mpg123] MPEG 1.0/2.0/2.5 layers I, II, III
**AUDIO: 8000 Hz, 2 ch, s16le, 8.0 kbit/3.12% (ratio: 1000->32000)**
Selected audio codec: [mpg123] afm: mpg123 (MPEG 1.0/2.0/2.5 layers I, II, III)

**AO: [pulse] 8000Hz 2ch s16le (2 bytes per sample)**
Video: no video
Starting playback...
A:   2.2 (02.1) of 3.0 (03.0)  0.1% 

***
With this config file
***

devices:
({
  type = "rtlsdr";
  index = 0;
  gain = 40;
  centerfreq = 462.500;
  afc = 1;
  channels:
  (
    {
      freq = 462.5625;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "1";
	  split_on_transmission = true;
        }
      );
    },
    {
      freq = 462.5875;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "2";
	  split_on_transmission = true;
        }
      );
    },
    {
      freq = 462.6125;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "3";
	  split_on_transmission = true;
        }
      );
    },
        {
      freq = 462.6375;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "4";
	  split_on_transmission = true;
        }
      );
    },
        {
      freq = 462.6625;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "5";
	  split_on_transmission = true;
        }
      );
    },
    {
      freq = 462.5875;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "6";
	  split_on_transmission = true;
        }
      );
    },
    {
      freq = 462.7125;
      modulation = "nfm";
      squelch_threshold = -15;
	outputs: (
        {
          type = "file";
          directory = ".";
	  #include_freq = true;
          filename_template = "7";
	  split_on_transmission = true;
        }
      );
    }
   );  
 }
);

**Comment by charlie-foxtrot** [comment]
I haven't had a chance to look at the code closer on the 8k vs 16k but will get back to you on that.

For opening a PR on the volume change, that would be great (please fork and open a PR to the `unstable` branch).  I suspect most of the code change will be on supporting the config parameter.  A `float` defaulting to 1.0 that is just multiplied to the WAV after demodulation [here-ish](https://github.com/charlie-foxtrot/RTLSDR-Airband/blob/main/src/rtl_airband.cpp#L625) should be all it takes, but the config file should support setting it per-channel (pre-mixer) and support multiple values for "scan-mode".  Take a look at `notch` as an example.
